### DISCUSSION TITLE
array-contains always returns an empty array on iOS

### DISCUSSION THREAD

**Discussion by klazbaba** [invertase/react-native-firebase#4956] [Q&A]
I have a query like this:

`const conversations = await firestore().collection('Conversations').where('type', '==', 'GROUP').where('participants', 'array-contains', getData('userDetails').id).get();
return conversations.docs;`

On Android, this returns the documents as expected, but on iOS it always returns an empty array.
Same thing happens when I try using array-contains-any.

**Comment by mikehardy** [comment]
Hi there! These are pretty heavily tested and I believe they run on android and ios each time we merge code, is there some test case we are missing?

https://github.com/invertase/react-native-firebase/blob/master/packages/firestore/e2e/Query/where.e2e.js

> mike@isabela:~/work/Invertase/react-native-firebase/packages/firestore/e2e (master) % grep -r array-contains

```
Query/where.e2e.js:  it('throws if query contains multiple array-contains', function () {
Query/where.e2e.js:        .where('foo.bar', 'array-contains', 123)
Query/where.e2e.js:        .where('foo.bar', 'array-contains', 123);
Query/where.e2e.js:      error.message.should.containEql('Queries only support a single array-contains filter');
Query/where.e2e.js:      firebase.firestore().collection(COLLECTION).where('foo.bar', 'array-contains');
Query/where.e2e.js:      firebase.firestore().collection(COLLECTION).where('foo.bar', 'array-contains', null);
Query/where.e2e.js:  it('throws if array-contains-any query with no array value', function () {
Query/where.e2e.js:      firebase.firestore().collection(COLLECTION).where('foo.bar', 'array-contains-any', '123');
Query/where.e2e.js:  it('throws if query has multiple array-contains-any filter', function () {
Query/where.e2e.js:        .where('foo.bar', 'array-contains-any', [1])
Query/where.e2e.js:        .where('foo.bar', 'array-contains-any', [2]);
Query/where.e2e.js:      error.message.should.containEql("You cannot use more than one 'array-contains-any' filter");
Query/where.e2e.js:  it('throws if query has array-contains-any & in filter', function () {
Query/where.e2e.js:        .where('foo.bar', 'array-contains-any', [1])
Query/where.e2e.js:        "You cannot use 'in' filters with 'array-contains-any' filters",
Query/where.e2e.js:  it('throws if query has in & array-contains-any filter', function () {
Query/where.e2e.js:        .where('foo.bar', 'array-contains-any', [2]);
Query/where.e2e.js:        "You cannot use 'array-contains-any' filters with 'in' filters",
Query/where.e2e.js:  it('returns with where array-contains filter', async function () {
Query/where.e2e.js:    const colRef = firebase.firestore().collection(`${COLLECTION}/filter/array-contains`);
Query/where.e2e.js:    const snapshot = await colRef.where('foo', 'array-contains', match.toString()).get();
Query/where.e2e.js:  it('returns with array-contains-any filter', async function () {
Query/where.e2e.js:      .collection(`${COLLECTION}/filter/array-contains-any${Date.now() + ''}`);
Query/where.e2e.js:    const snapshot = await colRef.where('category', 'array-contains-any', expect).get();
Query/where.e2e.js:  it('should correctly query integer values with array-contains operator', async function () {
Query/where.e2e.js:      .collection(`${COLLECTION}/filter/int-array-contains${Date.now() + ''}`);
Query/where.e2e.js:      .where('status', 'array-contains', 2)
Query/where.e2e.js:  it("should throw error when combining 'not-in' operator with 'array-contains-any' operator", async function () {
Query/where.e2e.js:      ref.where('test', 'array-contains-any', [2]).where('test', 'not-in', [1]);
```

> mike@isabela:~/work/Invertase/react-native-firebase/packages/firestore/e2e (master) % grep -r array-contains-any

```
Query/where.e2e.js:  it('throws if array-contains-any query with no array value', function () {
Query/where.e2e.js:      firebase.firestore().collection(COLLECTION).where('foo.bar', 'array-contains-any', '123');
Query/where.e2e.js:  it('throws if query has multiple array-contains-any filter', function () {
Query/where.e2e.js:        .where('foo.bar', 'array-contains-any', [1])
Query/where.e2e.js:        .where('foo.bar', 'array-contains-any', [2]);
Query/where.e2e.js:      error.message.should.containEql("You cannot use more than one 'array-contains-any' filter");
Query/where.e2e.js:  it('throws if query has array-contains-any & in filter', function () {
Query/where.e2e.js:        .where('foo.bar', 'array-contains-any', [1])
Query/where.e2e.js:        "You cannot use 'in' filters with 'array-contains-any' filters",
Query/where.e2e.js:  it('throws if query has in & array-contains-any filter', function () {
Query/where.e2e.js:        .where('foo.bar', 'array-contains-any', [2]);
Query/where.e2e.js:        "You cannot use 'array-contains-any' filters with 'in' filters",
Query/where.e2e.js:  it('returns with array-contains-any filter', async function () {
Query/where.e2e.js:      .collection(`${COLLECTION}/filter/array-contains-any${Date.now() + ''}`);
Query/where.e2e.js:    const snapshot = await colRef.where('category', 'array-contains-any', expect).get();
Query/where.e2e.js:  it("should throw error when combining 'not-in' operator with 'array-contains-any' operator", async function () {
Query/where.e2e.js:      ref.where('test', 'array-contains-any', [2]).where('test', 'not-in', [1]);
Query/where.e2e.js:        "You cannot use 'not-in' filters with 'array-contains-any' filters.",
```

**Reply by klazbaba** [reply]
Hello Mike,
Thanks for your response. These test cases seem pretty detailed.
I am not even getting any error thrown here. It just returns an empty array, as though nothing matches the query in the db.

**Reply by mikehardy** [reply]
> Hello Mike,
> Thanks for your response. These test cases seem pretty detailed.

They are - but to the point: is your case covered? Or is it missing?

> I am not even getting any error thrown here. It just returns an empty array, as though nothing matches the query in the db.

You indicate android works fine, is there any way you could test JS? Just curious on that one. The platform difference is pretty odd. You don't mention what version you are on, using dicusssions for issues means you don't include any information we need to help. Definitely suggest you use current version (10.8.1) without iOS SDK override so you get 7.6.0 of firebase-ios-sdk, and you might even try actually doing an override to get 7.7.0 (that's already passed tests internally here we just haven't released it yet - #4953 )

**Reply by klazbaba** [reply]
> > Hello Mike,
> > Thanks for your response. These test cases seem pretty detailed.
> 
> They are - but to the point: is your case covered? Or is it missing?
> 
> > I am not even getting any error thrown here. It just returns an empty array, as though nothing matches the query in the db.
> 
> You indicate android works fine, is there any way you could test JS? Just curious on that one. The platform difference is pretty odd. You don't mention what version you are on, using dicusssions for issues means you don't include any information we need to help. Definitely suggest you use current version (10.8.1) without iOS SDK override so you get 7.6.0 of firebase-ios-sdk, and you might even try actually doing an override to get 7.7.0 (that's already passed tests internally here we just haven't released it yet - #4953 )

I believe this part covers what I was expecting

it('returns with where array-contains filter', async function () {
const colRef = firebase.firestore().collection(`${COLLECTION}/filter/array-contains`);
const snapshot = await colRef.where('foo', 'array-contains', match.toString()).get();

And like I mentioned, it returns fine on Android. iOS also throws no exception at all. Only that it returns an empty array everytime.

**Reply by klazbaba** [reply]
> > Hello Mike,
> > Thanks for your response. These test cases seem pretty detailed.
> 
> They are - but to the point: is your case covered? Or is it missing?
> 
> > I am not even getting any error thrown here. It just returns an empty array, as though nothing matches the query in the db.
> 
> You indicate android works fine, is there any way you could test JS? Just curious on that one. The platform difference is pretty odd. You don't mention what version you are on, using dicusssions for issues means you don't include any information we need to help. Definitely suggest you use current version (10.8.1) without iOS SDK override so you get 7.6.0 of firebase-ios-sdk, and you might even try actually doing an override to get 7.7.0 (that's already passed tests internally here we just haven't released it yet - #4953 )

These are the current versions I am using:

"@react-native-firebase/app": "^10.8.1",
"@react-native-firebase/firestore": "^10.8.1",

As regards your suggestion on me testing in JS. How would you suggest I go about that please? Cos I am currently out of ideas myself.

**Reply by mikehardy** [reply]
Well, the general idea of react-native-firebase is that it implements the firebase-js-sdk exactly, where ever possible, just binding it down to native SDKs vs the JS SDK, and with minor differences in order to take advantage of the native platform when that is possible.

The general idea of troubleshooting difficult issues is that they should be reproducible with a minimally reproducible example - this actual becomes crucial when/if bugs are determined to be in the SDK, as the reproduction has to be self-contained and simple to quickly get attention and to verify fixes.

To combine the thoughts:
- creating an App.js that does nothing but create the required data fixture, then performs the queries and demonstrates the error is a critical effort. Then anyone can drop that into a clean project and see the problem
- once you have an App.js like this, it should *also* be possible to make a quick index.html page that uses firebase-js-sdk and the rough contents of your App.js in order to either verify that the queries really are fine and ios sdk likely has an issue, or not. This is just a possible avenue of investigation though, I am not certain it would advance things much beyond having a minimal repro

It's very suspicious that the test cases seem to cover your case (and they work? we think they work) but your project has a problem. I think pursuing a minimal reproduction is the best next step here - then once it is fully understood, perhaps adding a test case in the E2E tests for firestore that does the same, at which point we can start digging in
